diesel/src/mysql/connection/stmt/mod.rs

```diff
@@ -160,6 +160,11 @@ impl<'a> StatementUse<'a> {
             .map_err(|e| Error::DeserializationError(Box::new(e)))
     }
 
+    /// SAFETY:  This should only be called for INSERT queries.
+    pub(in crate::mysql::connection) unsafe fn insert_id(&self) -> u64 {
+        ffi::mysql_stmt_insert_id(self.inner.stmt.as_ptr())
```

Comment:
While that's how the documentation is written, I don't think that the result is UB.  The value is always present in the Ok packet and will always be a valid byte sequence for a u64.  The value absolutely might not make any sense if no `AUTO_INCREMENT` column is present, but that's not the same as the "spooky action at a difference" type of Undefined Behavior.

Looking at what `mysql_common` does, for example, the only check it performs is during packet deserialization; `last_insert_id` is an `Option<u64>` in their representation, and if `last_insert_id` on the wire is `0`, they consider that `None`

I also, FWIW, don't have any idea what kind of check we'd be able to perform in a query context for that

Reply:
I'm still wary to relay on something that's explicitly documented as not being defined as that could change at any point in the future. Remember that we want to provide a reasonable level of stability that should be decoupled from our dependencies. So yes, it might not be undefined behavior in terms of programming to read this value, the result still stays undefined and might change at any point. It's something that at least I do not feel comfortable with exposing in a stable API.

What we can easily do is to restrict it to insert statements with `Integer` primary keys at compile time by adding the relevant trait bounds on the target type for `InsertStatement`. What's harder to do is to check if the column is actually marked as `AUTO_INCREMENT`, as that would additional changes to include this information in the generated schema. That's still not impossible, although I think it might be to much trouble in this case.

Reply:
Another way to quickly get this feature is expose it under some unstable feature flag, if you not against that.

Reply:
I would rather prefer restricting it at least to Integer primary keys, as that isn't much harder than marking it as unstable feature. After all that should just require one or two lines of trait bounds on the relevant method.